File: test_regress/t/t_dist_getsetorder.pl
#!/usr/bin/env perl
if (!$::Driver) { use FindBin; exec("$FindBin::Bin/bootstrap.pl", @ARGV, $0); die; }
# DESCRIPTION: Verilator: Hacky import order checker, used to ensure all getters
# come before setters for consistent codegen when using autocxx (#5182)
#
# Copyright 2024 by Wilson Snyder. This program is free software; you
# can redistribute it and/or modify it under the terms of either the GNU
# Lesser General Public License Version 3 or the Perl Artistic License
# Version 2.0.
# SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0

scenarios(dist => 1);

my $root = "..";

### Must trim output before and after our file list
my %files = %{get_source_files($root)};

foreach my $file (sort keys %files) {
    my $filename = "$root/$file";
    next if !-f $filename;  # git file might be deleted but not yet staged
    next unless $file =~ /include.*verilated.*\.[ch].*$/;
    next if $file =~ /gtkwave/;

    my $contents = file_contents($filename);
    my %seen_setters;
    my %seen_getters;
    foreach my $line (split(/\n/, $contents . "\n\n")) {
        next if $line =~ /^\s*\/\//; # skip commented lines
        if ($line =~ /\s*void\s+([a-zA-Z0-9_]+)\([a-zA-Z0-9_]+.*/) {
            my $setter_name = $1;
            $seen_setters{$setter_name} = 1;
        } else {
            next unless $line =~ /\s*[a-zA-Z0-9_]+\s+([a-zA-Z0-9_]+)\(\s*\)/;
            my $getter_name = $1;
            if ($file =~ /verilated_sc_trace/ and $getter_name == "cycle") {
                next;  # hardcoded check for cycle() which looks like a setter but isn't
            }
            if (exists $seen_setters{$getter_name} and not (exists $seen_getters{$getter_name})) {
                error("$file: '$getter_name()' came after its setter; suggest swap order");
            }
            $seen_getters{$getter_name} = 1;
        }
    }
}

ok(1);
1;

sub get_source_files {
    my $root = shift;
    my $git_files = `cd $root && git ls-files`;
    print "MF $git_files\n" if $Self->{verbose};
    my %files;
    foreach my $file (split /\s+/, $git_files) {
        next if $file eq '';
        $files{$file} |= 1;
    }
    return \%files;
}

File: test_regress/t/t_dist_getsetorder.py
#!/usr/bin/env python3
# DESCRIPTION: Verilator: Hacky import order checker, used to ensure all getters
# come before setters for consistent codegen when using autocxx (#5182)
#
# Copyright 2024 by Wilson Snyder. This program is free software; you
# can redistribute it and/or modify it under the terms of either the GNU
# Lesser General Public License Version 3 or the Perl Artistic License
# Version 2.0.
# SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0

import vltest_bootstrap

test.scenarios('dist')

root = ".."


def get_source_files(root):
    git_files = test.run_capture("cd " + root + " && git ls-files")
    if test.verbose:
        print("MF " + git_files)
    files = {}
    for filename in git_files.split():
        if filename == '':
            continue
        files[filename] = True
    return files


if not os.path.exists(root + "/.git"):
    test.skip("Not in a git repository")

### Must trim output before and after our file list
files = get_source_files(root)

for filename in sorted(files.keys()):
    filename = os.path.join(root, filename)
    if not os.path.exists(filename):  # git file might be deleted but not yet staged
        continue
    if not re.search(r'include.*verilated.*\.[ch].*$', filename):
        continue
    if re.search(r'gtkwave', filename):
        continue

    contents = test.file_contents(filename) + "\n\n"
    seen_setters = {}
    seen_getters = {}
    lineno = 0
    for line in contents.splitlines():
        lineno += 1
        if re.search(r'^\s*\/\/', line):  # skip commented lines
            continue
        #print("L "+line)
        if re.search(r'^class', line):
            seen_setters = {}
            seen_getters = {}
            if test.verbose:
                print("C " + line)
            continue
        m = re.search(r'\s*void\s+([a-zA-Z0-9_]+)\([a-zA-Z0-9_]+.*', line)
        if m:
            setter_name = m.group(1)
            if test.verbose:
                print("S " + setter_name + " " + line)
            seen_setters[setter_name] = True
        m = re.search(r'\s*[a-zA-Z0-9_]+\s+([a-zA-Z0-9_]+)\(\s*\)', line)
        if m:
            getter_name = m.group(1)
            if getter_name == "cycle" and re.search(r'verilated_sc_trace', filename):
                continue  # hardcoded check for cycle() which looks like a setter but isn't
            if test.verbose:
                print("G " + setter_name + " " + line)
            if getter_name in seen_setters and getter_name not in seen_getters:
                test.error(filename + ":" + str(lineno) + ": '" + getter_name +
                           "()' came after its setter; swap order")
                seen_getters[getter_name] = True

test.passes()
